spec/support/geoip_mocking.rb

```diff
@@ -0,0 +1,205 @@
+# frozen_string_literal: true
+
+RSpec.configure do |config|
+  config.before(:each) do
+    full_description = RSpec.current_example&.full_description || ""
+    if full_description.include?("invalid UTF-8") || full_description.include?("underlying GEOIP has invalid")
+      allow(GEOIP).to receive(:city).and_return(
+        double(
+          country: double({ name: "Unit\xB7ed States", iso_code: "U\xB7S" }),
+          most_specific_subdivision: double({ iso_code: "C\xB7A" }),
+          city: double({ name: "San F\xB7rancisco" }),
+          postal: double({ code: "941\xB703" }),
+          location: double({ latitude: "103\xB7103", longitude: "103\xB7103" })
+        )
+      )
+    else
+      allow(GeoIp).to receive(:lookup) do |ip|
+        case ip
+        when "76.66.210.142"
+          GeoIp::Result.new(
+            country_name: "Canada",
+            country_code: "CA",
+            region_name: "ON",
+            city_name: "Toronto",
+            postal_code: "M5H 2N2",
+            latitude: nil,
+            longitude: nil
+          )
+        when "54.234.242.13"
+          GeoIp::Result.new(
+            country_name: "United States",
+            country_code: "US",
+            region_name: "VA",
+            city_name: "Ashburn",
+            postal_code: "20147",
+            latitude: nil,
+            longitude: nil
+          )
+        when "104.193.168.19"
+          GeoIp::Result.new(
+            country_name: "United States",
+            country_code: "US",
+            region_name: "CA",
+            city_name: "San Francisco",
+            postal_code: "94110",
+            latitude: nil,
+            longitude: nil
+          )
```

Comment:
wondering if we can maintain a hash with IP as key and GeoIP result as value. Can reduce the overall code lines?

Reply:
thanks! hash looks much more readable.